tests/test_generics.py

```diff
@@ -180,75 +208,159 @@ class Result(GenericModel, Generic[T]):
 
-    class Model(GenericModel, Generic[T, S]):
+    class Model(BaseModel, Generic[T, S]):
         x: T
         y: S
 
     with pytest.raises(TypeError) as exc_info:
         Model[int, int, int]
-    assert str(exc_info.value) == 'Too many parameters for Model; actual 3, expected 2'
 
-    with pytest.raises(TypeError) as exc_info:
-        Model[int]
-    assert str(exc_info.value) == 'Too few parameters for Model; actual 1, expected 2'
+    # This error message, which comes from `typing`, changed 'parameters' to 'arguments' in 3.11
+    error_message = str(exc_info.value)
+    assert error_message.startswith('Too many parameters') or error_message.startswith('Too many arguments')
+    assert error_message.endswith(
+        " for <class 'tests.test_generics.test_parameter_count.<locals>.Model'>; actual 3, expected 2"
+    )
 
 
-@pytest.mark.xfail(reason='working on V2')
 def test_cover_cache():
-    cache_size = len(_generic_types_cache)
+    cache_size = len(GENERIC_TYPES_CACHE)
+    T = TypeVar('T')
+
+    class Model(BaseModel, Generic[T]):
+        x: T
+
+    models = []  # keep references to models to get cache size
+
+    models.append(Model[int])  # adds both with-tuple and without-tuple version to cache
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 2
+    models.append(Model[int])  # uses the cache
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 2
+    del models
+
+
+def test_cache_keys_are_hashable():
+    cache_size = len(GENERIC_TYPES_CACHE)
+    T = TypeVar('T')
+    C = Callable[[str, Dict[str, Any]], Iterable[str]]
+
+    class MyGenericModel(BaseModel, Generic[T]):
+        t: T
+
+    # Callable's first params get converted to a list, which is not hashable.
+    # Make sure we can handle that special case
+    Simple = MyGenericModel[Callable[[int], str]]
+    models = []  # keep references to models to get cache size
+    models.append(Simple)
+
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 2
+    # Nested Callables
+    models.append(MyGenericModel[Callable[[C], Iterable[str]]])
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 4
+    models.append(MyGenericModel[Callable[[Simple], Iterable[int]]])
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 6
+    models.append(MyGenericModel[Callable[[MyGenericModel[C]], Iterable[int]]])
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 10
+
+    class Model(BaseModel):
+        x: MyGenericModel[Callable[[C], Iterable[str]]] = Field(...)
+
+    models.append(Model)
+    assert len(GENERIC_TYPES_CACHE) == cache_size + 10
+    del models
+
+
+def test_caches_get_cleaned_up():
+    types_cache_size = len(GENERIC_TYPES_CACHE)
+    T = TypeVar('T')
+
+    class MyGenericModel(BaseModel, Generic[T]):
+        x: T
+
+    Model = MyGenericModel[int]
+    assert len(GENERIC_TYPES_CACHE) == types_cache_size + 2
+    del Model
+    gc.collect()
+
+    # TODO: Need to figure out why this is failing
+    assert len(GENERIC_TYPES_CACHE) == types_cache_size
```

Comment:
It seems the WeakValuesDictionary approach to the types cache (from #5052) isn't evicting items as expected. Definitely would appreciate any help investigating this. @adriangb

Reply:
In #5052 we validated it by creating a ton of generics and measuring memory usage, not by checking the size of the cache. I'm a bit surprised by your tests behavior but not _too_ surprised, I don't know if it's easy to measure these things at this fine of a level.

I did a sanity check:

```python
import gc
from typing import Any, Generic, TypeVar
from weakref import WeakValueDictionary

T = TypeVar("T")

class MyGeneric(Generic[T]):
    pass

Concrete = MyGeneric[int]

cache: WeakValueDictionary[Any, Any] = WeakValueDictionary()

cache[(MyGeneric, int)] = Concrete

del Concrete

gc.collect(0)
gc.collect(1)
gc.collect(2)

assert len(cache) == 0  # actually 1
```

So it seems this would not work in the first place!

Reply:
Ok I think I found the issue:

```python
import gc
from typing import Any, Generic, TypeVar
from weakref import WeakValueDictionary

T = TypeVar("T")

class MyGeneric(Generic[T]):
    pass

cache: WeakValueDictionary[Any, Any] = WeakValueDictionary()

Concrete = MyGeneric[int]
cache[(MyGeneric, int)] = Concrete
del Concrete

gc.collect(0)
gc.collect(1)
gc.collect(2)

referrer = next(iter(gc.get_referrers(next(cache.values()))))
print(referrer)  # <functools._lru_cache_wrapper object at 0x102dee560>
print(referrer.__module__)  # typing
```

So it looks like this cache inside of `_tp_cache` of `typing.py` may be holding onto a reference.
It has a `maxsize` of 128, which basically means that the `WeakValueDict` will act as a regular dict for the _last_ 128 things that got run through it!

That explains why the test that creates thousands of objects and measures memory consumption looks like it's not accumulating but this one measuring just 1 reference does.

Reply:
Another little test:

```python
import gc
from typing import Any, Generic, TypeVar
from weakref import WeakValueDictionary
from pydantic import BaseModel

T = TypeVar("T")

class MyGeneric(BaseModel, Generic[T]):
    x: T

cache: WeakValueDictionary[Any, Any] = WeakValueDictionary()

def create_and_del() -> None:
    Concrete = MyGeneric[int]
    cache[(MyGeneric, int)] = Concrete
    del Concrete

create_and_del()
gc.collect(0)
gc.collect(1)
gc.collect(2)

for r in gc.get_referrers(next(cache.values())):
    print(r, end="\n\n")
```

```
(<class '__main__.MyGeneric[int]'>, <class '__main__.MyGeneric'>, <class 'pydantic.main.BaseModel'>, <class 'pydantic._internal._repr.Representation'>, <class 'typing.Generic'>, <class 'object'>)

{'type': 'model', 'cls': <class '__main__.MyGeneric[int]'>, 'schema': {'type': 'typed-dict', 'fields': {'x': {'schema': {'type': 'int', 'metadata': {}}, 'required': True, 'metadata': {'pydantic_json_schema_metadata': {'title': None, 'description': None, 'examples': None, 'extra_updates': None}}}}, 'return_fields_set': True, 'ref': '__main__.MyGeneric:5611558880[int:4315709880]'}, 'config': {'title': 'MyGeneric[int]', 'typed_dict_extra_behavior': 'ignore', 'allow_inf_nan': True, 'populate_by_name': False, 'str_strip_whitespace': False, 'str_to_lower': False, 'str_to_upper': False, 'strict': False, 'ser_json_timedelta': 'iso8601', 'ser_json_bytes': 'utf8'}, 'metadata': {'pydantic_json_schema_metadata': {'title': 'MyGeneric[int]', 'description': None}}}
```

The first one seems like the MRO? The second is a pydantic-core schema.

Reply:
> I'm a bit surprised by your tests behavior but not too surprised, I don't know if it's easy to measure these things at this fine of a level.

Just to be clear, I took this test directly from the v1 PR, so it should be working there.